### PR TITLE
Add buildinfo at compile time

### PR DESCRIPTION
**Pull request by loosebazooka**
Injects build info, which appears to be somewhat more portable that reading the jar manifest?

### CODE REVIEW

File: sigstore-java/build.gradle.kts [hunk 1]
```diff
@@ -120,3 +121,7 @@ jsonSchema2Pojo {
 tasks.named("sourcesJar") {
     dependsOn("generateJsonSchema2DataClassConfigRekor")
 }
+
+tasks.getByName("generateBuildInfo") {
+    (this as buildlogic.BuildInfoTask).packageName.set("dev.sigstore.buildinfo")
```
Comment (vlsi) [inline sigstore-java/build.gradle.kts:126]: Does the following work?

```suggestion
tasks.generateBuildInfo {
    packageName.set("dev.sigstore.buildinfo")
```

File: build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import buildlogic.BuildInfoTask
+import gradle.kotlin.dsl.accessors._8c47cae829ea3d03260d5ff13fb2398e.main
```
Comment (vlsi) [inline build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts:2]: Please remove this accidental import :)

```suggestion
```

**Comment by loosebazooka** [comment]
> Ideally, there should be a dependency from compileJava on generateBuildInfo so we know the file is properly generated before compile.

I believe this is implicit with the sourceSet
```
$ ~/src/sigstore-java$ ./gradlew sigstore-java:compileJava --dry-run
:sigstore-java:generateBuildInfo SKIPPED
:sigstore-java:extractProto SKIPPED
:sigstore-java:processResources SKIPPED
:sigstore-java:extractTestProto SKIPPED
:sigstore-java:processTestResources SKIPPED
:sigstore-java:generateJsonSchema2DataClassConfigRekor SKIPPED
:sigstore-java:generateJsonSchema2DataClass SKIPPED
:sigstore-java:extractIncludeProto SKIPPED
:sigstore-java:generateProto SKIPPED
:sigstore-java:compileJava SKIPPED
```

File: build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts (hunk 2 above)
Comment (loosebazooka) [reply build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts]: good catch, also had to switch from `java-base` to `java` to have access to the source sets

**Review by vlsi — APPROVED** [review]

File: build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import buildlogic.BuildInfoTask
+
+plugins {
+    `java`
+}
+
+val generateBuildInfoTask = tasks.register("generateBuildInfo", BuildInfoTask::class.java) {
```
[the remark this answers is not in the document]
Comment (loosebazooka) [reply build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts:7]: it doesn't need the task name anymore?

File: build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import buildlogic.BuildInfoTask
+
+plugins {
+    `java`
+}
+
+val generateBuildInfoTask by tasks.register("generateBuildInfo", BuildInfoTask::class.java) {
```
Comment (vlsi) [inline build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts:7]: It should be `registering`, not `register`

```suggestion
val generateBuildInfoTask by tasks.registering(BuildInfoTask::class) {
```

Comment (loosebazooka) [reply build-logic/jvm/src/main/kotlin/build-logic.build-info.gradle.kts:7]: ah, so I'll also have to change how it's resolved later in the sigstore-java/build.gradle.kts

**Review by vlsi — DISMISSED** [review]

File: build-logic/jvm/src/main/kotlin/buildlogic/BuildInfoTask.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    @TaskAction
+    fun run() {
+        val output = """
+            |package ${packageName.get()};
+            |
+            |public class BuildInfo {
+            |  public static final String VERSION = "${version.get()}";
+            |}
+        """.trimMargin()
```
[the remark this answers is not in the document]
Comment (loosebazooka) [reply build-logic/jvm/src/main/kotlin/buildlogic/BuildInfoTask.kt:32]: Sure. I have no preference.

**Review by vlsi — APPROVED** [review]

**Pull request merged**